### PR TITLE
Support macOS defaulting to spawn as the default multiprocessing method 

### LINKED ISSUE #55847: Python 3.8 and macOS switching default multi-processing to spawn
### Description of Issue
On macOS and under Python 3.8, the default multi-processing method switched from fork to spawn.

Salt is prepared for spawn but only for windows.
The salt code base needs to be updated to check the default multi-processing method instead of checking where(Windows, Linux, macOS) it's running in order to support non Windows platforms which default to spawn.

### Versions Report
All Salt versions up until Salt officially supports Python >= 3.8

### Links
https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
https://bugs.python.org/issue33725

### LINKED ISSUE #57742: [BUG] AttributeError: 'Process' object has no attribute '_args_for_getstate'
**Description**
salt-ssh crashes with an exception before doing seemingly anything

**Setup**
Just upgraded to salt 3001, 3000 was working fine

**Steps to Reproduce the behavior**
salt-ssh $host state.highstate test=True

<details><summary>AttributeError: 'Process' object has no attribute '_args_for_getstate'</summary>

```
Traceback (most recent call last):
  File "/usr/local/bin/salt-ssh", line 33, in <module>
    sys.exit(load_entry_point('salt==3001', 'console_scripts', 'salt-ssh')())
  File "/usr/local/Cellar/salt/3001/libexec/lib/python3.8/site-packages/salt/scripts.py", line 499, in salt_ssh
    client.run()
  File "/usr/local/Cellar/salt/3001/libexec/lib/python3.8/site-packages/salt/cli/ssh.py", line 26, in run
    ssh.run()
  File "/usr/local/Cellar/salt/3001/libexec/lib/python3.8/site-packages/salt/client/ssh/__init__.py", line 787, in run
    for ret in self.handle_ssh():
  File "/usr/local/Cellar/salt/3001/libexec/lib/python3.8/site-packages/salt/client/ssh/__init__.py", line 604, in handle_ssh
    routine.start()
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 283, in _Popen
    return Popen(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "/usr/local/Cellar/salt/3001/libexec/lib/python3.8/site-packages/salt/utils/process.py", line 755, in __getstate__
    args = self._args_for_getstate
AttributeError: 'Process' object has no attribute '_args_for_getstate'
```
</details>

**Expected behavior**
Salt executes on the host

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3001

Dependency Versions:
           cffi: 1.12.2
       cherrypy: unknown
       dateutil: 2.8.0
      docker-py: Not Installed
          gitdb: 2.0.6
      gitpython: 2.1.15
         Jinja2: 2.10.1
        libgit2: 1.0.1
       M2Crypto: 0.35.2
           Mako: 1.0.7
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: Not Installed
   pycryptodome: 3.9.7
         pygit2: 1.2.1
         Python: 3.8.3 (default, May 27 2020, 20:54:22)
   python-gnupg: 0.4.4
         PyYAML: 5.1.2
          PyZMQ: 19.0.0
          smmap: 3.0.4
        timelib: 0.2.4
        Tornado: 4.5.3
            ZMQ: 4.3.2

System Versions:
           dist: darwin 19.5.0
         locale: utf-8
        machine: x86_64
        release: 19.5.0
         system: Darwin
        version: 10.15.5 x86_64
```
</details>

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
Support macOS defaulting to spawn as the default multiprocessing method 

### What issues does this PR fix or reference?
We were previously forcing macOS to use fork instead of the the default spawn method, we fix that in this PR
Fixes https://github.com/saltstack/salt/issues/55847
Fixes https://github.com/saltstack/salt/issues/57742
Refs https://github.com/saltstack/salt/issues/61045

### Previous Behavior
We used the "fork" method for macOS

### New Behavior
We use the default spawn method for macOS

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog
- [x] Tests written/updated - The whole test suite under macOS

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by sagetherage** [comment]
fixes https://github.com/saltstack/salt/issues/55847

**Comment by weswhet** [comment]
on behalf of the macOS folks thank you, thank you, thank you!

**Comment by weswhet** [comment]
this looks great. I know you mentioned that this isn't likely to hit Sodium but it would be awesome if it could.

File: salt/utils/platform.py [hunk 1]
```diff
@@ -177,3 +178,13 @@ def is_fedora():
...
     ]
     return osname == "Fedora"
+
+
+@real_memoize
+def spanwing_platform():
+    """
+    Returns True if multiprocessing.get_start_method(allow_none=False) returns "spawn"
+
+    This is the default for Windows and macOS on Python >= 3.4
```
Comment (weswhet) [inline salt/utils/platform.py:188]: Per the python docs [here](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods). The default start method on macOS isn't changed to `spawn` until `3.8`

**Review by weswhet — COMMENTED** [review]
Building the python 3.7.4 macOS pkg with these changes still causes a crash when trying to multiprocess ObjC code.

File: salt/utils/platform.py (hunk 1 above)
Comment (s0undt3ch) [reply salt/utils/platform.py:188]: Fixed. Thanks!

**Comment by s0undt3ch** [comment]
> Building the python 3.7.4 macOS pkg with these changes still causes a crash when trying to multiprocess ObjC code.
Should we force "spawn" on 3.7 then?

**Comment by s0undt3ch** [comment]
Made some updates but it's not yet ready.
There's still a lot :/ that needs to change...

**Comment by weswhet** [comment]
> Made some updates but it's not yet ready.
> There's still a lot :/ that needs to change...

You must have read my mind :) I was in the process of pulling the new changes to test. Let me know when you think it's ready for a test drive! Thanks!

**Comment by sagetherage** [comment]
@s0undt3ch looks like you have merge conflicts to resolve here, pretty please.

**Comment by s0undt3ch** [comment]
Yes, and the work to get this over the finish line is immense.
Unfortunately, I don't see this going in for Mg :/

File: salt/utils/process.py [hunk 2]
```diff
@@ -703,35 +698,32 @@ def kill_children(self, *args, **kwargs):
 
 
 class Process(multiprocessing.Process, NewStyleClassMixIn):
+    def __new__(cls, *args, **kwargs):
+        # We implement __new__ because we want to capture the passed in *args and **kwargs
```
Comment (dwoz) [inline salt/utils/process.py:702]: I'm still not a fan of this :) If feels like we are trying to be magic for no reason. I don't see the problem with having `__getstate__` and `__setstate__` explicitly defined where it's being used. It's more obvious what's going on that way. IMO.

**Review by dwoz — CHANGES_REQUESTED** [review]

File: salt/utils/process.py (hunk 2 above)
Comment (s0undt3ch) [reply salt/utils/process.py:702]: I know, and I'm still a fan of it.
It assures we, or anyone, never forgets, and removed duplicate boilerplate code to support pickling

Comment (s0undt3ch) [reply salt/utils/process.py:702]: Also, we're already using a custom `multiprocessing.Process` class which alleviates on required boilerplate code, so, why is adding additional support to remove boilerplate code such an issue?

**Comment by natewalck** [comment]
Hey all! I work on the CPE team @ Square and I recently updated a few open source projects to use universal Python builds in order to support the new M1 macOS devices. 

@weswhet Let me know that this was a blocker for using Python 3.9.1 and I was curious if there was anything I could help with on this. 

If Salt can get 3.9.1 support, I'm pretty confident I can get the Universal build working for M1 support (3.9.1 was the first to support this).

Thanks for all your work on this!

**Comment by weswhet** [comment]
😍

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
@weswhet do you get notifications on pushes? 😃

**Comment by weswhet** [comment]
> @weswhet do you get notifications on pushes? 😃

Haha I do, well for PR's I like 😄

How's it going?

**Comment by s0undt3ch** [comment]
Just picked this again, right now I only solved conflicts and created myself a Catalina VM to run tests locally.

**Comment by weswhet** [comment]
Cool let me know if you need anything! @natewalck and I are more than happy to help in any way.

**Comment by s0undt3ch** [comment]
@weswhet, @natewalck: I will need to set this PR aside for a while.
Would one of you pick it up and get it over the finish line?

The current status is getting the test suite to pass on macOS. It fails because there are still places where we check for `salt.utils.platform.is_windows()` where we should now be checking for `salt.utils.platform.spawning_platform()`

**Comment by natewalck** [comment]
> @weswhet, @natewalck: I will need to set this PR aside for a while.
> Would one of you pick it up and get it over the finish line?
> 
> The current status is getting the test suite to pass on macOS. It fails because there are still places where we check for `salt.utils.platform.is_windows()` where we should now be checking for `salt.utils.platform.spawning_platform()`

I assume this still needs to be finished, correct? I cloned down your branch and I am going to take a look at it.

**Comment by s0undt3ch** [comment]
@natewalck Thank You!

**Comment by natewalck** [comment]
This is going to be a lot of "fun" I think. Running the tests on a M1 machine, it looks like some of the libs are not universal:

```
nox -e 'pytest-3(coverage=True)'
nox > Running session pytest-3(coverage=True)
nox > Re-using existing virtual environment at .nox/pytest-3-coverage-true.
nox > Session pytest-3(coverage=True) was successful.
nox > Running session pytest-parametrized-3(crypto=None, transport='zeromq', coverage=True)
nox > Re-using existing virtual environment at .nox/pytest-parametrized-3-crypto-none-transport-zeromq-coverage-true.
nox > python -m pip install --progress-bar=off -U pip setuptools!=50.*,!=51.*,!=52.*
nox > python -m pip install --progress-bar=off -r requirements/static/ci/py3.9/darwin.txt

nox > python -m pip install --progress-bar=off coverage==5.2
nox > coverage erase
nox > python -m coverage run -m pytest --showlocals --rootdir /Users/nw/code/salt --log-file=artifacts/logs/runtests-20210729150910.169754.log --log-file-level=debug --show-capture=no -ra -s --transport=zeromq
ImportError while loading conftest '/Users/nw/code/salt/tests/conftest.py'.
tests/conftest.py:40: in <module>
    from tests.support.helpers import (
tests/support/helpers.py:46: in <module>
    from tests.support.sminion import create_sminion
tests/support/sminion.py:17: in <module>
    import salt.minion
salt/minion.py:24: in <module>
    import salt.crypt
salt/crypt.py:31: in <module>
    import salt.utils.rsax931
salt/utils/rsax931.py:136: in <module>
    libcrypto = _init_libcrypto()
salt/utils/rsax931.py:89: in _init_libcrypto
    libcrypto = _load_libcrypto()
salt/utils/rsax931.py:82: in _load_libcrypto
    return cdll.LoadLibrary(_find_libcrypto())
/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py:452: in LoadLibrary
    return self._dlltype(name)
/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py:374: in __init__
    self._handle = _dlopen(self._name, mode)
E   OSError: dlopen(/opt/salt/lib/libcrypto.dylib, 6): no suitable image found.  Did find:
E   	/opt/salt/lib/libcrypto.dylib: mach-o, but wrong architecture
E   	/opt/salt/lib/libcrypto.dylib: mach-o, but wrong architecture
nox > Command python -m coverage run -m pytest --showlocals --rootdir /Users/nw/code/salt --log-file=artifacts/logs/runtests-20210729150910.169754.log --log-file-level=debug --show-capture=no -ra -s --transport=zeromq failed with exit code 4
```

**Comment by weswhet** [comment]
@s0undt3ch the changes you have for `tests/unit/utils/test_thin.py` cause the test to fall. It runs fine on master without the changes. 

I'm running `nox -e 'pytest-zeromq-3(coverage=False)' -- tests/unit/utils/test_thin.py -x`

getting an output of 

```
================================================================================= ERRORS ==================================================================================
_____________________________________________________________ ERROR collecting tests/unit/utils/test_thin.py ______________________________________________________________
tests/unit/utils/test_thin.py:41: in <module>
    class SSHThinTestCase(TestCase):
tests/unit/utils/test_thin.py:148: in SSHThinTestCase
    ???
tests/support/runtests.py:136: in __getattribute__
    return object.__getattribute__(self, name)
E   AttributeError: 'RuntimeVars' object has no attribute 'PYTEST_SESSION'
============================================================================ warnings summary =============================================================================
salt/ext/tornado/web.py:1805
  /Users/duotest2/dev/soundsalt/salt/ext/tornado/web.py:1805: DeprecationWarning: invalid escape sequence \.
    """A collection of request handlers that make up a web application.

-- Docs: https://docs.pytest.org/en/stable/warnings.html
========================================================================= short test summary info =========================================================================
ERROR tests/unit/utils/test_thin.py - AttributeError: 'RuntimeVars' object has no attribute 'PYTEST_SESSION'
```

Let me know what you think.

**Comment by s0undt3ch** [comment]
@weswhet yeah, that variable no longer exists.
I will try to carve some time tomorrow to rebase this PR against latest master.

How does that sound? Am I going to break anything you're working on?

**Comment by weswhet** [comment]
> @weswhet yeah, that variable no longer exists.
> I will try to carve some time tomorrow to rebase this PR against latest master.
> 
> How does that sound? Am I going to break anything you're working on?

Nope, that sounds great. @natewalck and I are working on this, this week. Want to give us access to your fork to push changes or what's the best way to go about us contributing here?

**Comment by weswhet** [comment]
@s0undt3ch Can you resend the invitation to edit, it expired before I could accept and I've got some changes to push.

**Comment by s0undt3ch** [comment]
@weswhet set. Also, I rebased earlier today, hopefully it doesn't mess your changes ...

**Comment by weswhet** [comment]
@s0undt3ch all good thanks! I'm having a really tough time getting any integration test to run at all. I can upload some logs in a bit but unit tests are running well so I'm working my way through those.

**Comment by Ch3LL** [comment]
I would say that we need to be able to run the test suite on all supported python versions, so it should work on both 3.7 and 3.9. When we add big sur tests after this fix is merged, then we will be also testing 3.9 on bigsur. But @s0undt3ch can correct me if I'm wrong here.

**Comment by s0undt3ch** [comment]
I think the pickle exceptions is what needs to be fixed, regardless of the python version, but Wes was looking into this last, so he might know more.
@bryceml  did mention that we were going to try and get a separate Mac VM(not big sur) with 3.9 installed, don't know the status of that.

**Comment by Ch3LL** [comment]
@weswhet are you still blocked?

**Comment by weswhet** [comment]
@Ch3LL pretty much. The pickle exceptions are a little out of my league. @s0undt3ch is this something you can look into? I can further look after the full test suite once we have jobs running py3.9. Until then I think I'm pretty much stuck, unfortunately.

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by weswhet** [comment]
@s0undt3ch I'll be taking some time off work next month so if there is anything I can help with let me know. Looks like your getting there! All us macOS folks really appreciate it!

**Comment by s0undt3ch** [comment]
Yup, getting close, getting close.
Just try to take the test suite for some spins and see if you can figure out the fixes :)

Yeah easier said than done.

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Comment by s0undt3ch** [comment]
re-run full all

**Review by twangboy — APPROVED** [review]

**Review by Ch3LL — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Comment by s0undt3ch** [comment]
🎆

**Comment by natewalck** [comment]
You are all heroes <3

**Pull request merged**